smarts/env/wrappers/metrics.py

```diff
@@ -0,0 +1,273 @@
+# Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import copy
+import functools
+from dataclasses import dataclass, fields
+from typing import Any, Dict, Set, TypeVar
+
+import gym
+
+from smarts.env.wrappers.metric import termination
+from smarts.env.wrappers.metric.costs import CostFuncs, Costs
+from smarts.env.wrappers.metric.counts import Counts
+
+_MAX_STEPS = 800
+
+
+@dataclass
+class Record:
+    """Stores an agent's performance count and cost values."""
+
+    counts: Counts
+    costs: Costs
+
+
+@dataclass
+class Data:
+    """Stores an agent's performance record and cost functions."""
+
+    record: Record
+    cost_funcs: CostFuncs
+
+
+class Metrics(gym.Wrapper):
+    """Metrics class wraps an underlying _Metrics class. The underlying
+    _Metrics class computes agents' performance metrics in a SMARTS
+    environment. Whereas, this Metrics class is a basic gym.Wrapper class
+    which prevents external users from accessing or modifying attributes
+    beginning with an underscore, to ensure security of the metrics computed.
+
+    Args:
+        env (gym.Env): A gym.Env to be wrapped.
+
+    Raises:
+        AttributeError: Upon accessing an attribute beginning with an underscore.
+
+    Returns:
+        gym.Env: A wrapped gym.Env which computes agents' performance metrics.
+    """
+
+    def __init__(self, env: gym.Env):
+        env = _Metrics(env)
```

Comment:
If the _Metrics utility is wrapped like this, how do we get at score and records?

The `gym` way to get this information might be `env.render("score")` perhaps?

Reply:
We still can get the score and records by calling `env.score()` and `env.records()`.

Reply:
Ah, right, it forwards through `__getattr__`.